File: frontend/lib/slots.ts
import type { FullBuild } from "@/types/api";

/**
 * Même ordre que `app/solver/slots.py` (backend).
 * Les 6 emplacements dofus1–6 acceptent aussi bien les Dofus que les Trophées.
 */
export const SLOT_DEFS = [
  { id: "hat", label: "Chapeau" },
  { id: "cloak", label: "Cape" },
  { id: "amulet", label: "Amulette" },
  { id: "ring1", label: "Anneau 1" },
  { id: "ring2", label: "Anneau 2" },
  { id: "belt", label: "Ceinture" },
  { id: "boots", label: "Bottes" },
  { id: "weapon", label: "Arme" },
  { id: "shield", label: "Bouclier" },
  { id: "dofus1", label: "Dofus / Trophée 1" },
  { id: "dofus2", label: "Dofus / Trophée 2" },
  { id: "dofus3", label: "Dofus / Trophée 3" },
  { id: "dofus4", label: "Dofus / Trophée 4" },
  { id: "dofus5", label: "Dofus / Trophée 5" },
  { id: "dofus6", label: "Dofus / Trophée 6" },
  { id: "pet", label: "Familier" },
] as const;

export type SlotId = (typeof SLOT_DEFS)[number]["id"];

export function emptyBuild(): Record<SlotId, number | null> {
  return Object.fromEntries(SLOT_DEFS.map((s) => [s.id, null])) as Record<
    SlotId,
    number | null
  >;
}

export function mergeFullBuildSlots(fb: FullBuild): Record<SlotId, number | null> {
  const base = emptyBuild();
  for (const s of SLOT_DEFS) {
    const v = fb.slots[s.id];
    base[s.id] = v === undefined || v === null ? null : v;
  }
  return base;
}

File: backend/app/solver/slots.py
"""Equipment slot definitions (`type_name_id` = canonical slug from Ankama `type.id`, see `etl/equipment_type_ids.py`)."""

from __future__ import annotations

from app.models.item import Item

# Slots used in the MIP (one item chosen per slot).
SLOT_ORDER: list[str] = [
    "hat",
    "cloak",
    "amulet",
    "ring1",
    "ring2",
    "belt",
    "boots",
    "weapon",
    "shield",
    "dofus1",
    "dofus2",
    "dofus3",
    "dofus4",
    "dofus5",
    "dofus6",
    "pet",
]

# Anciens noms acceptés en entrée (migration/rétrocompatibilité).
LEGACY_SLOT_ALIASES: dict[str, str] = {
    "trophy1": "dofus5",
    "trophy2": "dofus6",
}


def _excluded(item: Item) -> bool:
    t = item.type_name_id or ""
    if "certificate" in t or t.startswith("perceptor-"):
        return True
    if t in {"tool", "sidekick", "prysmaradite"}:
        return True
    return False


def item_fits_slot(slot: str, item: Item) -> bool:
    if _excluded(item):
        return False
    t = item.type_name_id
    if not t and slot not in ("weapon",):
        return False
    if slot in ("hat", "cloak", "amulet", "belt", "boots"):
        return t == slot
    if slot in ("ring1", "ring2"):
        return t == "ring"
    if slot == "weapon":
        return bool(item.is_weapon)
    if slot == "shield":
        return t == "shield"
    if slot.startswith("dofus"):
        # Les 6 emplacements dofus acceptent les dofus ET les trophées.
        return t in ("dofus", "trophy")
    if slot == "pet":
        return t in ("pet", "mount")
    return False
